extensions-core/namespace-lookup/src/main/java/io/druid/query/lookup/NamespaceLookupExtractorFactory.java

```diff
@@ -0,0 +1,276 @@
+      baos.write(StringUtils.toUtf8(NamespaceLookupExtractorFactory.class.getCanonicalName()));
+      baos.write(0xFF);
+      CLASS_CACHE_KEY = baos.toByteArray();
+    }
+    catch (IOException e) {
+      // Should never happen
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @JsonIgnore
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  @JsonIgnore
+  private final ReadWriteLock startStopSync = new ReentrantReadWriteLock();
+  @JsonIgnore
+  private final NamespaceExtractionCacheManager manager;
+  @JsonIgnore
+  private final LookupIntrospectHandler lookupIntrospectHandler;
+  private final ExtractionNamespace extractionNamespace;
+  private final long firstCacheTimeout;
+  private final boolean oneToOne;
+
+  private final String extractorID;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace") ExtractionNamespace extractionNamespace,
+      @JsonProperty("firstCacheTimeout") Long firstCacheTimeout,
+      @JsonProperty("oneToOne") boolean oneToOne,
+      @JacksonInject NamespaceExtractionCacheManager manager
+  )
+  {
+    this.extractionNamespace = Preconditions.checkNotNull(
+        extractionNamespace,
+        "extractionNamespace should be specified"
+    );
+    this.firstCacheTimeout = firstCacheTimeout == null ? DEFAULT_SCHEDULE_TIMEOUT : firstCacheTimeout;
+    Preconditions.checkArgument(this.firstCacheTimeout >= 0);
+    this.oneToOne = oneToOne;
+    this.manager = manager;
+    this.extractorID = buildID();
+    this.lookupIntrospectHandler = new LookupIntrospectHandler()
+    {
+      @GET
+      @Path("/keys")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getKeys()
+      {
+        return Response.ok(getLatest().keySet().toString()).build();
+      }
+
+      @GET
+      @Path("/values")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getValues()
+      {
+        return Response.ok(getLatest().values().toString()).build();
+      }
+
+      @GET
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getMap()
+      {
+        return Response.ok(getLatest()).build();
+      }
+
+      private Map<String, String> getLatest()
+      {
+        return ((MapLookupExtractor) get()).getMap();
+      }
+    };
+  }
+
+  @VisibleForTesting
+  public NamespaceLookupExtractorFactory(
+      ExtractionNamespace extractionNamespace,
+      NamespaceExtractionCacheManager manager
+  )
+  {
+    this(extractionNamespace, null, false, manager);
+  }
+
+  @Override
+  public boolean start()
+  {
+    final Lock writeLock = startStopSync.writeLock();
+    writeLock.lock();
+    try {
+      if (!started.compareAndSet(false, true)) {
+        LOG.warn("Already started! [%s]", extractorID);
+        return true;
+      }
+      if (!manager.scheduleAndWait(extractorID, extractionNamespace, firstCacheTimeout)) {
+        LOG.warn("Failed to schedule lookup [%s]", extractorID);
```

Comment:
No, it might also be due to failure in scheduled thing that is to run. For example, if you gave it a config that doesn't find or doesn't parse the specified location. It will retry up to the time limit, then return false if it never succeeded in the allotted time frame